fastestimator/op/tensorop/model/update.py

```diff
@@ -74,22 +107,130 @@ def fe_retain_graph(self, retain: Optional[bool] = None) -> Optional[bool]:
+
+        if self.framework == "tf":
+            # scale up loss for mixed precision training to avoid underflow
+            if self.model.mixed_precision:
+                loss = self.model.current_optimizer.get_scaled_loss(loss)
+            # for multi-gpu training, the gradient will be combined by sum, normalize the loss
+            strategy = tf.distribute.get_strategy()
+            if isinstance(strategy, tf.distribute.MirroredStrategy):
+                loss = loss / strategy.num_replicas_in_sync
+
+        elif self.framework == "torch":
+            # scale up loss for mixed precision training to avoid underflow
+            if self.model.current_optimizer.scaler is not None:
+                loss = self.model.current_optimizer.scaler.scale(loss)
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return loss
+
+    def _get_gradient(self, loss, state):
+        if self.framework == "tf":
+            gradients = get_gradient(loss, self.model.trainable_variables, tape=state['tape'])
+
+        elif self.framework == "torch":
+            trainable_params = [p for p in self.model.parameters() if p.requires_grad]
+            try:
+                gradients = get_gradient(loss, trainable_params, retain_graph=self.retain_graph)
+            except RuntimeError as err:
+                if err.args and isinstance(err.args[0], str) and err.args[0].startswith(
+                        'one of the variables needed for gradient computation has been modified by an inplace operation'
+                ):
+                    raise RuntimeError(
+                        "When computing gradients for '{}', some variables it relied on during the forward pass had"
+                        " been updated. Consider setting defer=True in earlier UpdateOps related to models which "
+                        "interact with this one.".format(self.model.model_name))
+                raise err
+
+        else:
+            raise ValueError(f"Unrecognized framework {self.framework}")
+
+        return gradients
+
+    def _gradient_postprocess(self, gradients, state):
+        """ Gradient postprocess for multi-GPU and mixed-precision training
+        """
+        if self.framework == "tf":
+            if self.gradients is not None:  # user provide gradients
+                strategy = tf.distribute.get_strategy()  # need tape.stop_recording() ?
+                # for multi-gpu training, the gradient will be combined by sum, normalize the gradient
+                if isinstance(strategy, tf.distribute.MirroredStrategy):
+                    gradients = gradients / strategy.num_replicas_in_sync
+
+            # scale down gradient to balance scale-up loss
+            if self.model.mixed_precision:
+                with state["tape"].stop_recording():
+                    gradients = self.model.current_optimizer.get_unscaled_gradients(gradients)
+
+        elif self.framework == "torch":
```

Comment:
if we move the framework guard to .build we can eliminate the elif / else here

Reply:
fixed